app_dart/lib/src/model/appengine/branch.dart

```diff
@@ -0,0 +1,66 @@
+// Copyright 2021 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:cocoon_service/src/model/appengine/key_converter.dart';
+import 'package:gcloud/db.dart';
+import 'package:github/github.dart';
+import 'package:json_annotation/json_annotation.dart';
+
+part 'branch.g.dart';
+
+@Kind(name: 'Branch', idType: IdType.String)
+class Branch extends Model<String> {
+  Branch({Key<String>? key, required int lastActivity}) {
+    parentKey = key?.parent;
```

Comment:
There's no parents for `Branch`, so this is dead code

Reply:
umm maybe I am wrong, but my understanding is it if we keep the code this way, when we initialize the key using `final Key<String> key = datastore.db.emptyKey.append<String>(Branch, id: id);`, the parentKey will be set to `datastore.db.emptyKey` , which is a bit different from a `null` parentKey (what we would obtain if we remove this line).

When we are looking up and deleting entities, and providing a `currentBranch.key`, we are calling on the getter function of `Key<T> get key => parentKey!.append(runtimeType, id: id);` (which is not modifiable) part of the code to retrieve the key. This line of code asserts on parentKey not being `null`. If we initialized the value to be `datastore.db.emptyKey` we can pass this check. If we remove this line, parentKey will be defaulted to `null`, and causing the assertion of `parentKey!` to fail when we lookup or delete entities. 

My understanding is based on how we currently generate and lookup keys. Maybe there is a way to workaround this. I could be wrong but I thought with the current topology we need this line to properly initialize a non null parentKey.

Reply:
What you're describing is an issue in the tests. We should update the fake datastore to handle null parent keys

Reply:
umm so I have the logic at L115 - L120 of githubwebhook.dart currently looking like this https://github.com/flutter/cocoon/pull/1688/files#diff-3a0003489aad220cc50ef5c753a82ed536839ae6ed0ada1ac8b45c12e3ad1f2aR115-R123 . Maybe there is a workaround for this part but I am not sure how to use null as a base key for `append` to work

Reply:
The linked pasted doesn't work.

Can you share the snippet?

Reply:
For sure, it looks like 
```    final String id = '$repository/$branch';
    final Key<String> key = datastore.db.emptyKey.append<String>(Branch, id: id);
    final Branch currentBranch = Branch(key: key, lastActivity: lastActivity);
    try {
      await datastore.lookupByValue<Branch>(currentBranch.key);
    } on KeyNotFoundException {
      await datastore.insert(<Branch>[currentBranch]);
    }
```

Reply:
I'm not sure what the error here is. That's expected, and I don't see issues with it. There's no need to save the parent key as it's always going to be empty/null

Reply:
My understanding is that when we are calling `currentBranch.key`, we are calling on the getter method `Key<T> get key => parentKey!.append(runtimeType, id: id);`. In this getter method, it first creates a parentKey, assert its not null (but could be empty), and then append. So if we have a null value for parentKey it will fail, but an empty key is different from null in that it is a Key instance

Reply:
one of the sample errors looks like this 
```:00:03 +139 -1: test/request_handlers/github_webhook_test.dart: github webhook create branch event should add branch to db if db is empty [E]
  Null check operator used on a null value
  package:gcloud/src/db/models.dart 106:30                               Model.key
  package:cocoon_service/src/service/branch_service.dart 42:59           BranchService.handleCreateRequest
  ===== asynchronous gap ===========================
  package:cocoon_service/src/request_handlers/github_webhook.dart 88:11  GithubWebhook.post
  ===== asynchronous gap ===========================
  test/request_handlers/github_webhook_test.dart 2053:7                  main.<fn>.<fn>
```

Reply:
the line 106 it points to is from the key model class which is not modifiable
```
abstract class Model<T> {
  T? id;
  Key? parentKey;

  Key<T> get key => parentKey!.append(runtimeType, id: id);       //this is the line 106
}
```

Reply:
SGTM